core/rio/turtle/src/main/java/org/eclipse/rdf4j/rio/turtle/TurtleWriter.java

```diff
@@ -256,57 +276,71 @@ protected void handleStatementInternal(Statement st, boolean endRDFCalled, boole
 		Resource subj = st.getSubject();
 		IRI pred = st.getPredicate();
 		Value obj = st.getObject();
+
 		try {
-			if (subj.equals(lastWrittenSubject)) {
-				if (pred.equals(lastWrittenPredicate)) {
-					// Identical subject and predicate
-					writer.write(",");
-					wrapLine(prettyPrint);
-				} else {
-					// Identical subject, new predicate
-					writer.write(";");
-					writer.writeEOL();
-
-					// Write new predicate
-					writer.decreaseIndentation();
-					writePredicate(pred);
-					writer.increaseIndentation();
-					wrapLine(true);
-					path.removeLast();
-					path.addLast(pred);
-					lastWrittenPredicate = pred;
-				}
+			if (inlineBNodes && (pred.equals(RDF.FIRST) || pred.equals(RDF.REST))) {
+				handleList(st);
```

Comment:
List handling in the pretty printed turtle code has always been bad. I should make some tests for that to see if you've fixed it now.

Reply:
None of what I've done really touches the list handling I think - the only thing that changes is how it is invoked.